Add "Save Snippet" sidebar button + display custom inner content in the sidebar

sobo-odoo opened the pull request.

robodoo commented:
This PR targets the un-managed branch odoo-dev/odoo:master-mysterious-egg, it needs to be retargeted before it can be merged.

robodoo commented:
This PR targets the un-managed branch odoo-dev/odoo:master-mysterious-egg, it needs to be retargeted before it can be merged.

FrancoisGe commented on addons/html_builder/static/src/builder/builder_sidebar/tabs/block_tab/block_tab.js:
```diff
@@ -105,4 +105,9 @@ export class BlockTab extends Component {
     }
+
+    renameCustomInnerContent(ev, snippet) {
+        const newName = ev.currentTarget.parentElement.querySelector("input")?.value;
```
input ? What is the use case ?

FrancoisGe commented on addons/html_builder/static/tests/custom_tab/container_buttons.test.js:
```diff
@@ -57,6 +59,34 @@ test("Use the sidebar 'clone' buttons", async () => {
+    expect(saveButtonSelector).toHaveCount(1);
+
+    // TODO improve when "request_save" will be done.
+    // Maybe make a tour to test the behavior ?
```
What is the behavior ?

FrancoisGe commented on addons/html_builder/static/src/builder/builder_sidebar/builder_sidebar.js:
```diff
@@ -150,10 +151,17 @@ export class BuilderSidebar extends Component {
             this.env.services
         );
 
+        this.context = Object.assign({}, user.context, {
```
you don t need user.context we send it by default

FrancoisGe commented on addons/html_builder/static/src/builder/builder_sidebar/builder_sidebar.js:
```diff
@@ -150,10 +151,17 @@ export class BuilderSidebar extends Component {
         );
 
+        this.context = Object.assign({}, user.context, {
+            website_id: this.websiteService.currentWebsite.id,
+            lang: this.websiteService.currentWebsite.metadata.lang,
+            user_lang: user.context.lang,
```
maybe use the for load snippet ?

FrancoisGe commented on addons/html_builder/static/src/builder/snippet_model.js:
```diff
@@ -181,4 +205,69 @@ export class SnippetModel extends Reactive {
+                body: _t(
+                    "To save a snippet, we need to save all your previous modifications and reload the page."
+                ),
+                cancel: () => resolve(),
```
onClose: resolve,

FrancoisGe commented on addons/html_builder/static/src/builder/snippet_model.js:
```diff
@@ -181,4 +205,69 @@ export class SnippetModel extends Reactive {
+                        // HTML, otherwise the popup will be saved invisible and
+                        // therefore not visible in the "add snippet" dialog.
+                        snippetCopyEl = snippetEl.cloneNode(true);
+                    }
```
It looks not very good ?

FrancoisGe commented on addons/html_builder/static/src/builder/snippet_model.js:
```diff
@@ -181,4 +205,69 @@ export class SnippetModel extends Reactive {
+                    const defaultSnippetName = isButton
+                        ? _t("Custom Button")
+                        : _t("Custom %s", snippetEl.dataset.name);
+                    snippetCopyEl = snippetCopyEl || snippetEl.cloneNode(true);
```
No diff with snippetCopyEl ?

FrancoisGe commented on addons/html_builder/static/src/builder/snippet_model.js:
```diff
@@ -181,4 +205,69 @@ export class SnippetModel extends Reactive {
+                        : _t("Custom %s", snippetEl.dataset.name);
+                    snippetCopyEl = snippetCopyEl || snippetEl.cloneNode(true);
+                    snippetCopyEl.classList.add("s_custom_snippet");
+                    delete snippetCopyEl.dataset.name;
```
no name ?

sobo-odoo replied:
The `save_snippet` orm call is in charge of putting a new name (it checks which ones are available and so on) :slightly_smiling_face:

FrancoisGe commented on addons/html_builder/static/src/builder/snippet_model.js:
```diff
@@ -181,4 +205,69 @@ export class SnippetModel extends Reactive {
+                            editableParentEl = editableEl;
+                            break;
+                        }
+                    }
```
Why check each editable ? We should use closest(editableSelector) ?

FrancoisGe commented on addons/html_builder/static/src/builder/snippet_model.js:
```diff
@@ -181,4 +205,69 @@ export class SnippetModel extends Reactive {
+                        }
+                    }
+
+                    this.context["model"] = editableParentEl.dataset.oeModel;
```
Not a good idea to edit the object directly

const context = {...this.context, model: ....}

FrancoisGe commented on addons/html_builder/static/src/builder/snippet_model.js:
```diff
@@ -181,4 +205,69 @@ export class SnippetModel extends Reactive {
+                    this.context["model"] = editableParentEl.dataset.oeModel;
+                    this.context["field"] = editableParentEl.dataset.oeField;
+                    this.context["resId"] = editableParentEl.dataset.oeId;
+                    await rpc("/web/dataset/call_kw/ir.ui.view/save_snippet", {
```
why not use orm.call like delete_snippet ?

FrancoisGe commented on addons/html_builder/static/src/builder/snippet_model.js:
```diff
@@ -181,4 +205,69 @@ export class SnippetModel extends Reactive {
+                            context: this.context,
+                        },
+                    });
+                    resolve();
```
add new customSnippet to snippetsByCategory.snippet_custom. It will trigger the reactivity. We probably not need to reload the page

sobo-odoo replied:
In the end, I simply reloaded the snippets by calling `load`, because I still need to get the snippet html after the call (in order to be able to drop it afterwards), which is not easy to retrieve from the orm call
So the easiest was to just reload the snippets :slightly_smiling_face:

FrancoisGe commented on addons/html_builder/static/src/builder/builder_sidebar/tabs/block_tab/block_tab.xml:
```diff
@@ -17,7 +17,41 @@
+                <button class="fa fa-trash btn o_we_hover_danger"
+                        t-attf-data-tooltip="Delete {{snippet.title}}"
+                        t-on-click="() => this.props.snippetModel.deleteCustomSnippet(snippet)"/>
+            </div>
```
maybe create a small custom comp ?

FrancoisGe commented on addons/html_builder/static/src/builder/builder_sidebar/tabs/block_tab/block_tab.js:
```diff
@@ -105,4 +105,9 @@ export class BlockTab extends Component {
+
+    renameCustomInnerContent(ev, snippet) {
+        const newName = ev.currentTarget.parentElement.querySelector("input")?.value;
+        this.props.snippetModel.renameCustomSnippet(snippet, newName);
```
do you know why we reload something ? Block ui isnt cool

sobo-odoo commented:
I made the changes (see [diff](https://github.com/odoo-dev/odoo/compare/b2480c88fb22e94a736f43044b957cc2b9929b10..12bd409e7706a28a46e8adc44074e8687e6fff39) + [other diff](https://github.com/odoo-dev/odoo/compare/12bd409e7706a28a46e8adc44074e8687e6fff39..c4f9a5780be896ce656dd5e9ed190e2fc8730ab9) improving the rest of the of the test) :slightly_smiling_face:

FrancoisGe commented on addons/html_builder/static/src/builder/snippet_model.js:
```diff
@@ -181,4 +202,74 @@ export class SnippetModel extends Reactive {
+                        // its children in the case of a popup, or it will be
+                        // saved as invisible and will therefore not be visible
+                        // in the "add snippet" dialog).
+                        const revertOperation = makeSavePoint();
+                        const rootEl = snippetEl.matches(".s_popup")
+                            ? snippetEl
+                            : snippetEl.parentNode;
+                        cleanForSaveHandlers.forEach((handler) => handler({ root: rootEl }));
```
use deep clone (remove makesavepoint)

FrancoisGe commented on addons/html_builder/static/src/builder/builder_sidebar/tabs/block_tab/block_tab.xml:
```diff
@@ -33,4 +40,34 @@
     </div>
 </t>
 
+<t t-name="html_builder.CustomInnerSnippet">
```
move it to his own file same for js

FrancoisGe commented on addons/html_builder/static/src/builder/builder_sidebar/tabs/block_tab/block_tab.xml:
```diff
@@ -33,4 +40,34 @@
+                    <input t-ref="rename-input" type="text" autocomplete="chrome-off" t-att-value="snippet.title" class="text-start" t-on-pointerdown="(ev) => ev.stopPropagation()"/>
+                    <button class="o_we_text_success fa fa-check btn btn-outline-success border-0"
+                            data-tooltip="Confirm"
+                            t-on-click="() => this.onConfirmRename()"/>
```
t-on-click="onConfirmRename"

FrancoisGe commented on addons/html_builder/static/src/builder/builder_sidebar/tabs/block_tab/block_tab.xml:
```diff
@@ -33,4 +40,34 @@
+        <div t-if="!state.isRenaming" class="rename-delete-buttons float-end">
+            <button class="fa fa-pencil btn o_we_hover_success btn-outline-info border-0"
+                    t-attf-data-tooltip="Rename {{snippet.title}}"
+                    t-on-click.stop="() => state.isRenaming = true"/>
```
I prefer a small function to switch the state

FrancoisGe commented on addons/html_builder/static/src/builder/builder_sidebar/tabs/block_tab/block_tab.xml:
```diff
@@ -33,4 +40,34 @@
+            <div class="thumbnail-img" t-attf-style="background-image: url({{snippet.thumbnailSrc}});"/>
+            <t t-if="state.isRenaming">
+                <div class="rename-input w-100 mx-1">
+                    <input t-ref="rename-input" type="text" autocomplete="chrome-off" t-att-value="snippet.title" class="text-start" t-on-pointerdown="(ev) => ev.stopPropagation()"/>
```
t-on-pointerdown.stop="() => {}" (Why you need to do it ?)

sobo-odoo replied:
Because the sidebar or the container (don't remember which one) does something on pointerDown, which means that every time I clicked on the input, this click was canceled by this something, so I could not write in it 
But I think this is temporary, when we will make the style,... of the sidebar correctly, maybe this could be removed ? I will see at that time :+1:

FrancoisGe commented on addons/html_builder/static/src/builder/builder_sidebar/tabs/block_tab/block_tab.xml:
```diff
@@ -33,4 +40,34 @@
+                            t-on-click="() => this.onConfirmRename()"/>
+                    <button class="o_we_text_danger fa fa-times btn btn-outline-danger border-0"
+                            data-tooltip="Cancel"
+                            t-on-click="() => state.isRenaming = false"/>
```
same

FrancoisGe commented on addons/html_builder/static/src/website_builder_action.scss:
```diff
@@ -54,3 +58,7 @@
+
+body:has(.o_builder_sidebar_open) .o_notification_manager {
+    @include o-position-absolute($top: map-get($spacers, 2), $right: calc(#{$o-we-sidebar-width} + 0.5rem));
+}
```
space

sobo-odoo replied:
this is because of prettier :smiling_face_with_tear: I cannot add it back :grimacing:

sobo-odoo commented:
I made the changes (see [diff](https://github.com/odoo-dev/odoo/compare/0cfd31b8f68dd69d62e5a0cc006fa7303627a72c..935584f8e9d631369f2f21e585b73717804db1c7)) :slightly_smiling_face: 
I only clean for save the clone now (so I removed the savepoint) and so I fixed the clean_for_save_handlers in order to be consistent and to take into account the root element too (first commit) :slightly_smiling_face:
